Proposed blog post: "Cut-over logic in vitess schema migrations" (#1001)
vitessio/website: content/en/blog/2022-03-30-online-ddl-vitess-cut-over.md

Vitess migrations are:

- Blocking & atomic to vitess clients
- Safe from data loss / data drift concerns

To understand better what this means, let's first review what cut-over is.

## The cut-over step

This is the most critical step in an online schema migration process, and in all implementations and tools: from [pt-online-schema-change](https://www.percona.com/doc/percona-toolkit/3.0/pt-online-schema-change.html), [fb-osc](http://bazaar.launchpad.net/~mysqlatfacebook/mysqlatfacebook/tools/annotate/head:/osc/OnlineSchemaChange.php), to [gh-ost](https://github.com/github/gh-ost) and finally [vitess](https://vitess.io/docs/user-guides/schema-changes/).

> An online schema migration (aka online DDL) process creates a ghost table, populates it, brings it up to date with the original table, and then cuts-over by renaming the original table away, and the ghost table in its place.

Reviewer: Do we want to use `ghost table` or `shadow table`?

Author: `shadow` table it is

The cut-over is a point where the two tables must be, and remain, in complete sync throughout the switch. Whichever the implementation is, this involves locks, and this impacts the users/apps.

> Vitess migrations use asynchronous change propagation: they look for changes to the original table in the binary logs, then apply those changes to the ghost table. This means some latency, however small, before applying those changes to the shadow table. At any point in time the two tables may be out of sync, with the ghost table slightly lagging behind the original table.

Reviewer: I see we are using `shadow table` at least once. We should pick one and use that throughout. FWIW, I vote for `shadow` over `ghost`.

Author: `shadow` table it is

VTTablet has the notion of ACLs (Access Control Lists). These were primarily intended to let administrators deny writes to tables.

Each query that goes through VTTablet gets a query plan. The plan includes all ACLs associated with any table referenced in the query. At execution time, VTTablet first evaluates whether ACLs permit the query to pass, and then proceeds to execute the query on MySQL and return its results.

Vitess migrations introduce a new form of ACL: time limited buffering. These are expiring rules. You may set a buffering rule onto some table, then either cancel the rule, or it eventually expires by itself.

While the rule is active, a query is buffered, or essentially just kept at bay, waiting until further notice. Either:

- the rule will be actively cancelled, meaning buffering is done, and the query proceeds to execute (assuming no other ACL has conflicts), or
- the rule will self expire, in which case the query gets rejected.

> We expect cut-over total time to be a few seconds, typically two or three. When cut-over begins, vitess sets a `10s` buffering ACL on the migrated table. Thus, any new query enters buffering for up to ten seconds. If migration completes by then, great, the query unblocks and proceeds to run on the new table. If not, then the query errors.

Reviewer: Just out of curiosity, is the `10s` something that is hard-coded into the migration cutover process?

Author: Yes, right now it is: https://github.com/vitessio/vitess/blob/0e7879b0f27d26ea9fc684d66ea23ba30e6b3d33/go/vt/vttablet/tabletserver/query_executor.go#L66

> Buffering can only allow as many queries to wait. Under heavy load the app could eventually exhaust the buffer capacity, at which time queries will error. This is why it is essential to complete the cut-over as soon as possible.

Reviewer: ```suggestion
Buffering can only allow a certain (configurable) number of queries to wait. Under heavy load the app could eventually exhaust the buffer capacity, at which time queries will error. This is why it is essential to complete the cut-over as soon as possible.
```

However, buffering is not a complete solution:

1. What happens to a query which was executed _just before_ cut-over, and has validated its ACLs? If we place buffering ACLs now, it's too late for that query. It proceeds to execute. How can we tell?
2. What happens if some automation runs a direct query on the MySQL server? It's not the normal vitess flow, but it can happen, and we've all been there.

### 2. Stall

> As a gesture to potential pending queries, the flow stalls for an extra `100ms`. That time is very likely to let any queries which already passed pre-cut-over ACLs to begin executing on the MySQL server, also likely to let them complete executing.

Reviewer: ```suggestion
As a gesture to potential pending queries, the flow stalls for an extra `100ms`. That time is very likely enough to let any queries which already passed pre-cut-over ACLs to begin executing on the MySQL server, also likely enough to let them complete executing.
```

Author: Updated per suggestion

Vitess migrations use asynchronous change propagation: they look for changes to the original table in the binary logs, then apply those changes to the ghost table. This means some latency, however small, before applying those changes to the shadow table. At any point in time the two tables may be out of sync, with the ghost table slightly lagging behind the original table.

But at cut-over, the ghost table must be brought into complete sync with the original table. How is this done?

## A brief suspense

All schema migration techniques use some form of locking at cut-over time. That locking causes queries to stall and connections to spike.

To be able to bring the ghost table up to speed with the original table, vitess runs the following generalized flow:

- Verify the migration is in appropriate state
- Prevent writes to the original table
- Mark the point in time where writes have been disabled
- Consume binary logs up to marked point in time, and apply onto ghost table
- Tables now in sync, cut-over

> Exactly ow vitess prevents write to the original table is the topic of the reminder of this post. It's noteworthy that the flow can fail, or timeout, in which case vitess resumes writes and tries again at a later stage.

Reviewer 2: ```suggestion
Exactly how vitess prevents write to the original table is the topic of the reminder of this post. It's noteworthy that the flow can fail, or timeout, in which case vitess resumes writes and tries again at a later stage.
```

Author: Thank you! Fixed

`gh-ost`'s approach is to lock the table via an [elaborate mechanism](https://github.com/github/gh-ost/issues/82) that ensures rollback in case of error/timeout. To the users/apps it looks atomic. Queries will block and pile up, then be released to operate on the new table. The logic relies on internal MySQL lock prioritization, and some users have shown specific scenarios where expected prioritization [can fail](https://github.com/github/gh-ost/issues/887).

Implementing vitess's cut-over, we wanted the best of all worlds. We wanted the cut-over to appear atomic to the apps, i.e. the apps should see no unexpected errors, and at worst case will block for a few seconds. We also wanted complete certainty of data integrity: that no write can possibly take place on the original table while we bring the ghost table to speed. Breakdown follows.

### 1. Buffering writes

> Vitess has the advantage that traffic goes through VTGate, its MySQL compatible proxy. Normal uses and apps do not communicate directly to MySQL (though as explained below, the cut-over logic also covers direct communication scenarios). Queries are normally sent to VTGate, which routes them to the appropriate shards and to the VTTablet servers on those shards. VTTablet will then run the query on MySQL.

Reviewer 2: ```suggestion
Vitess has the advantage that traffic goes through VTGate, its MySQL compatible proxy. Normal users and apps do not communicate directly to MySQL (though as explained below, the cut-over logic also covers direct communication scenarios). Queries are normally sent to VTGate, which routes them to the appropriate shards and to the VTTablet servers on those shards. VTTablet will then run the query on MySQL.
```
?

Author: Thank you! Fixed

Each query that goes through VTTablet gets a query plan. The plan includes all ACLs associated with any table referenced in the query. At execution time, VTTablet first evaluates whether ACLs permit the query to pass, and then proceeds to execute the query on MySQL and return its results.

Vitess migrations introduce a new form of ACL: time limited buffering. These are expiring rules. You may set a buffering rule onto some table, then either cancel the rule, or it eventually expires by itself.

While the rule is active, a query is buffered, or essentially just kept at bay, waiting until further notice. Either:

- the rule will be actively cancelled, meaning buffering is done, and the query proceeds to execute (assuming no other ACL has conflicts), or
- the rule will self expire, in which case the query gets rejected.

We expect cut-over total time to be a few seconds, typically two or three. When cut-over begins, vitess sets a `10s` buffering ACL on the migrated table. Thus, any new query enters buffering for up to ten seconds. If migration completes by then, great, the query unblocks and proceeds to run on the new table. If not, then the query errors.

> Buffering can only allow as many queries to wait. Each query runs on its own connection. As more queries come in, more connections will pile up. Under heavy load the app could eventually exhaust the connections limit, at which time queries will error. This is why it is essential to complete the cut-over as soon as possible.

Reviewer 3: > Buffering can only allow as many queries to wait.

How many queries can it allow to wait? As many as there are available connections?

Author: Yes indeed. And to be honest I don't have that number.  _hurriedly reviewing code_

Reviewer 4: Buffering can allow more queries than available connections! That's the point of buffering. It'll queue them until the connections free up. The default number of connections to buffer, and the max duration for buffering, are customizable options, unrelated to the connection pool.

Author: @vmg right! I was wondering about `vtgate` connections capacity, and it seems like there is no specific cap for that.

Reviewer: That's correct. The buffer capacity might be set to some not-very-high number like `1000`.
So we should say something like 
```
Under heavy load the app could eventually exhaust the buffer capacity
```
and maybe delete this part
```
Each query runs on its own connection. As more queries come in, more connections will pile up. 
```

- Failure in renaming the original table away? No problem, undo ACLs and try later
- Failure in renaming the ghost table in place of the original table? No problem, rename the original table back in place, remove buffering, try again later

What happens if VTTablet's process fails while the puncture is in place, though? This is where we see the vitess framework benefit. A new VTTablet will run. Whether we failover to a new MySQL server or not, we expect there to eventually be a VTTablet process in charge. That process will run recovery steps for prematurely interrupted migrations. It will in fact resume any interrupted migration from point of interruption.

Before renaming the original table away, VTTablet audits the intended action. In case of failure, the new VTTablet processes the state of interrupted migrations and sees that audit. It then restores the original table back in place, thus rolling back the entire cut-over operation. ACLs are in-memory and so the new VTTablet does not need to remove the buffering ACL. It then adopts the migration and lets it run, and in the natural order of things attempts to cut-over when appropriate.

## Summary

> By taking advantage of the vitess framework itself, vitess migration are able to deliver a multi layered cut-over mechanism, involving ACLs as well as MySQL primitives, such that users and apps get the best experience, while still maintaining complete control over data accuracy.

Reviewer 3: ```suggestion
By taking advantage of the vitess framework itself, vitess migrations are able to deliver a multi layered cut-over mechanism, involving ACLs as well as MySQL primitives, such that users and apps get the best experience, while still maintaining complete control over data accuracy.
```

Author: Thank you! Fixed.
